src/core/composer/qgscomposerpointsbasedshape.cpp

```diff
@@ -0,0 +1,404 @@
+    QgsComposition* c )
+    : QgsComposerItem( c ),
+    mTagName( tagName ),
+    mSelectedPoint( -1 ),
+    mDrawPoints( false )
+{
+  const QRectF boundingRect = polygon.boundingRect();
+  setSceneRect( boundingRect );
+
+  // change polygon coordinate in bounding area reference system
+  const QPointF topLeft = boundingRect.topLeft();
+
+  QVector<QPointF>::iterator it = polygon.begin();
+  for ( ; it != polygon.end(); ++it )
+  {
+    QPointF nextPt = *it;
+    nextPt.setX( nextPt.x() - topLeft.x() );
+    nextPt.setY( nextPt.y() - topLeft.y() );
+    mPolygon.append( nextPt );
+  }
+}
+
+QgsComposerPointsBasedShape::~QgsComposerPointsBasedShape()
+{
+}
+
+double QgsComposerPointsBasedShape::computeDistance( const QPointF &pt1,
+    const QPointF &pt2 ) const
+{
+  return sqrt( pow( pt1.x() - pt2.x(), 2 ) + pow( pt1.y() - pt2.y(), 2 ) );
+}
+
+bool QgsComposerPointsBasedShape::addPoint( const QPointF &pt,
+    const bool checkArea,
+    const double radius )
+{
+  const QPointF start = convertToItemCoordinate( pt );
+  double min_distance = std::numeric_limits<double>::max();
+  double max_distance = ( checkArea ) ? radius : min_distance;
+  bool rc = false;
+  int idx = -1;
+
+  for ( int i = 0; i != mPolygon.size(); i++ )
+  {
+    // get points of polyline
+    const QPointF pt1 = mPolygon[i];
+    QPointF pt2 = mPolygon[0];
+    if (( i + 1 ) != mPolygon.size() )
+      pt2 = mPolygon[i+1];
+
+    // compute line eq
+    const double coef = ( pt2.y() - pt1.y() ) / ( pt2.x() - pt1.x() );
+    const double b = pt1.y() - coef * pt1.x();
+
+    double distance = std::numeric_limits<double>::max();
+    if ( isinf( coef ) )
+      distance = qAbs( pt1.x() - start.x() );
+    else
+    {
+      const double coef2 = ( -1 / coef );
+      const double b2 = start.y() - coef2 * start.x();
+
+      QPointF inter;
+      if ( isinf( coef2 ) )
+      {
+        distance = qAbs( pt1.y() - start.y() );
+        inter.setX( start.x() );
+        inter.setY( pt1.y() );
+      }
+      else
+      {
+        const double interx = ( b - b2 ) / ( coef2 - coef );
+        const double intery = interx * coef2 + b2;
+        inter.setX( interx );
+        inter.setY( intery );
+      }
+
+      // check if intersection is within the line
+      const double length1 = computeDistance( inter, pt1 );
+      const double length2 = computeDistance( inter, pt2 );
+      const double length3 = computeDistance( pt1, pt2 );
+      const double length4 = length1 + length2;
+
+      if ( qAbs( length3 - length4 ) < std::numeric_limits<float>::epsilon() )
+        distance = computeDistance( inter, start );
+    }
+
+    if ( distance < min_distance && distance < max_distance )
+    {
+      min_distance = distance;
+      idx = i;
+    }
+  }
+
+  if ( idx >= 0 )
+  {
+    rc = _addPoint( idx, start, max_distance );
+    updateSceneRect();
+  }
+
+  return rc;
+}
+
+QPointF QgsComposerPointsBasedShape::convertToItemCoordinate( const QPointF &point )
```

Comment:
@mhugo 

> it's not clear for me here. Passing by value would be better because the function returns a QPointF and then copy ellision will occur ? So passing by value saves a copy of reference ? Is it what you had in mind ?

I'm going off advice given by the KDAB guys (who know MUCH more about optimising qt code then me ;) ). Apparently for QPoint/QPointF/QSize/QSizeF given that the size of the object itself is comparable or smaller than a 64 bit pointer and the copy constructor is so trivial it's actually a lot more efficient to copy those classes then to pass a reference and have the cost of dereferencing it. It's almost certainly over-optimising for this function, but I think it's good practice to get into and I've already converted the rest of the codebase to pass these classes by value.

But yes, it also simplifies the function and there's already a new object being constructed here in any case.